### ISSUE TITLE
Undefined symbol: calloc in >=2.0.29 thin LTO

### ISSUE THREAD

**Issue by curiousdannii** [emscripten-core/emscripten#15676]
I just tried updating the Emscripten version used by [Emglken](https://github.com/curiousdannii/emglken) from 2.0.27 to 2.0.32, and am now getting two undefined symbol errors, for calloc and __syscall_getgroups32. I tried 3.0.0 and it didn't help either.

```
error: undefined symbol: calloc (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: _calloc may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library

error: undefined symbol: __syscall_getgroups32 (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: ___syscall_getgroups32 may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
```

I tried adding them to EXPORTED_FUNCTIONS, and it wasn't fixed. I didn't expect it would be.

So I think the __syscall_getgroups32 error is from this change in 2.0.32:

I'm guessing this is from this change:

>Stub functions from library_syscall.js and library.js were replaced with native code stubs (See system/lib/libc/emscripten_syscall_stubs.c). This should be better for wasm module portability as well as code size. As part of this change the return value of popen was fixed to return NULL rather than -1 and the getpwnam family of functions were changed to return an error rather than throw a JavaScript exception (this behaviour matches what the other stub functions do). As before, the ALLOW_UNIMPLEMENTED_SYSCALLS setting controls whether of not these stubs get included at link time, and STRICT disables this setting.

After disabling STRICT and enabling ALLOW_UNIMPLEMENTED_SYSCALLS that fixes the __syscall_getgroups32 error, but not the calloc error. Just setting ALLOW_UNIMPLEMENTED_SYSCALLS as well as STRICT doesn't work, I had to disable STRICT. Am I missing much from doing so?

The calloc error starts occurring in 2.0.29, about which the changelog has no details. Great! There are only 9 commits between 2.0.28 and 2.0.29, and none look relevant to me: https://github.com/emscripten-core/emscripten/compare/2.0.28...2.0.29

Any ideas?

**Comment by sbc100** [comment]
It should be enough to enable `-sALLOW_UNIMPLEMENTED_SYSCALLS`.   I verified that `-sSTRICT` + `-sALLOW_UNIMPLEMENTED_SYSCALLS` does what it supposed to do (at least with `__syscall_mincore`).

The calloc issue llooks different, can you share your full emcc command linker command line?

**Comment by sbc100** [comment]
Can you try building with `-sLLD_REPORT_UNDEFINED` on the off change it reports better error about `calloc` (I somehow doubt it will, but its worth a try).

**Comment by curiousdannii** [comment]
```
/emsdk/upstream/emscripten/emcc -O3 -DNDEBUG -flto=thin  -sALLOW_MEMORY_GROWTH=1 -sASYNCIFY=1 -sASYNCIFY_IGNORE_INDIRECT=1 "-sASYNCIFY_REMOVE=['gli_get_*','glk_get_*']" -sENVIRONMENT=web -sEXIT_RUNTIME=1 "-sEXPORTED_FUNCTIONS=['_main','_gidispatch_get_game_id']" "-sEXPORTED_RUNTIME_METHODS=['AsciiToString','FS']" -sMODULARIZE=1 -sSTRICT=1 -sWASM=1 -Wl,--wrap=getc,--wrap=ungetc --minify 0 -O3 -sTEXTDECODER=2 @CMakeFiles/glulxe-core.dir/objects1.rsp  -o glulxe-core.js @CMakeFiles/glulxe-core.dir/linklibs.rsp
```

With `-sLLD_REPORT_UNDEFINED` I get this:

```
wasm-ld: error: /emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/thinlto/libdlmalloc.a(dlmalloc.o): undefined symbol: calloc
wasm-ld: error: /emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/thinlto/libdlmalloc.a(dlmalloc.o): undefined symbol: calloc
wasm-ld: error: /emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/thinlto/libdlmalloc.a(dlmalloc.o): undefined symbol: calloc
```

**Comment by sbc100** [comment]
Can you try removing `-flto=thin`?    `calloc` should literally be defined in `dlmalloc.o` but you could be running into some strange LTO issue.

**Comment by curiousdannii** [comment]
Disabling LTO does fix it.

**Comment by sbc100** [comment]
You can try full LTO.  We have very little coverage of thin LTO so its likely to have issue.  See #15664

**Comment by curiousdannii** [comment]
I wasn't sure how to change LTO modes in CMake, as there's [no setting for that yet](https://gitlab.kitware.com/cmake/cmake/-/issues/22913), but this seems to work, and it solves the undefined symbol error:

```
#[[set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)]]
set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -flto=full")
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -flto=full")
```

If thin LTO is risky, could Emscripten's CMake setup be changed to use full LTO instead?

**Comment by sbc100** [comment]
I'm trying to make sure the thinLTO works in all cases.. can you confirm if your project still has an issue with thinLTO (with the latest 3.1.5 release)?

If so, then it would be great if we could try to build a reproducer.. since non of test are showing this issue.

**Comment by tralves** [comment]
Upgrading to 3.1.5 fixed the exact same issue with `-flto` for me. Cheers!

**Comment by curiousdannii** [comment]
@sbc100 Sorry for taking so long. I've just tried updating to Emscripten 3.1.8 and changing back to thinLTO, but it's still broken. 

This is my project: https://github.com/curiousdannii/emglken/tree/thinlto
`npm run build` to build it.

**Comment by sbc100** [comment]
I think this should be fixed by #17866.

**Comment by curiousdannii** [comment]
Yes it's now fixed in 3.1.22!
